src/util/tokenpipe.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#include <util/tokenpipe.h>
+
+#include <config/bitcoin-config.h>
+
+#ifndef WIN32
+
+#include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+std::optional<TokenPipe> TokenPipe::create()
+{
+    int fds[2] = {-1, -1};
+#if HAVE_O_CLOEXEC && HAVE_DECL_PIPE2
+    if (pipe2(fds, O_CLOEXEC) != 0) {
+        return std::nullopt;
+    }
+#else
+    if (pipe(fds) != 0) {
+        return std::nullopt;
+    }
+#endif
+    return TokenPipe(fds);
+}
+
+TokenPipe::~TokenPipe()
+{
+    Close();
+}
+
+void TokenPipe::Close()
+{
+    if (m_fds[0] != -1) close(m_fds[0]);
+    if (m_fds[1] != -1) close(m_fds[1]);
```

Comment:
I think this would be incorrect if you had code like: `{ TokenPipe tp; a(); tp.Close(); b(); }` in particular -- the fds would be closed when `Close()` is called, but may be reallocated to different files during `b()` which would then be incorrectly closed when `tp` is destructed.

Could explicitly set `m_fds = {-1,-1}`, or could drop `Close()` entirely and move the code into the destructor -- it's only called from `operator=(TokenPipe&&)` and you could just mark that as `= deleted;` I think; it doesn't seem to be needed.

Reply:
Will take a look. Agree the fds should be se to -1 after closing them to prevent 'use fd after close' problems, seems I forgot that. I made it like this to be symmetric with `TokenPipeEnd::Close`. Sure, I could move the code around if minimalism was the goal, but I like a consistent API more.